### ISSUE TITLE
poor mixing for one parameter

### ISSUE THREAD

**Issue by fkrauer** [itsdfish/DifferentialEvolutionMCMC.jl#39]
Hi @itsdfish 

I have been toying with the DEzs MCMC now for a while. Initially I had poor mixing for all model parameters, but I tuned the sampler parameters a bit and achieved acceptable (albeit not perfect) chain mixing for all of them except one. The results below are achieved with `burnin=250000, n_iter=300000, n_groups=3, Np=3, θsnooker=0.1, burnin, α=0.2, β=0.2, ϵ=0.001, σ=0.1, κ=0.5`. They are also not fully converged, so may need to run longer. I will continue to improve the mixing also for ρ₀, but do you have any tips in the meanwhile? Could it be that the mixing of ρ₀ is so poor because its scale is much smaller than the other params? TIA

![chain_16](https://user-images.githubusercontent.com/36159611/121325167-aaaca400-c911-11eb-9dfa-cb1efb76a9e7.png)

**Comment by itsdfish** [comment]
Your suspicion is probably right. In my experience, this algorithm is sensitive to the scaling of the parameters. For example,  converge decreases if you change the scaling of the data in this [example](https://github.com/itsdfish/DifferentialEvolutionMCMC.jl/blob/7ec6ef6b465bcbf0ed943e9f03b622c740436c6d/Examples/Multivariate_Guassian_Example.jl#L8) to

```
μs = rand(Normal(0.0, .0001), n_μ)
data = rand(MvNormal(μs, .0001), n_d)
```

The NUTS sampler has the same problem. One solution is to reparametrize the model so that it is in a different scale. The Stan [documentation](https://mc-stan.org/docs/2_18/stan-users-guide/reparameterization-section.html) is a good resource for rescaling. 

Another potential factor might be the sensitivity of the log likelihood to a change in the parameter. For example,  the MCMC sampler may struggle to converge if the log likelihood does not change much for some parameters, but does change a lot for others. This might be the case for kappa and phi vs the other parameters.

**Comment by fkrauer** [comment]
I have changed the parametrization a bit for some params, but it didn't really solve the problem yet. I may have to add some scaling factors or something alike. I also noticed that in the first few iterations, the mixing is strong (as it should be in my opinion), but then the chains become stationary very quickly. This is with `θsnooker=0.1, α=0.1, β=0.1, ϵ=0.001, σ=0.05, κ=0.1` but occurs also if `κ=1.0`, it is always the first 50 iterations. Which sampler parameter influences this? I should note that I use burnin=0. 
![mixing](https://user-images.githubusercontent.com/36159611/121687614-d9b64780-cac2-11eb-80a6-e60579ec228d.png)

**Comment by itsdfish** [comment]
That is really odd. 

This is probably not the issue, but I recommend setting a burnin > 0 with the default `update_particle!` function, which is `random_gamma`. During burnin, this update function places more weight on particles near the mode. Typically this improves sampling efficiency.  You can retain the burnin samples with `discard_burnin = false`. This may or may not help in this case.

I have been wondering whether three particles is sufficient in all cases. Perhaps it is, but the paper only tested this in a small set of conditions. It might be worth increasing the number of particles. 

You probably have `sample=resampler` specified, but I just want to make sure. If that is set to the default sample function, it will generally not work well with only 3 particles. 

If none of those recommendations work, it might be worth verifying that the likelihood function is working as intended. One way I check is with the following procedure:

1. select a set of parameters
2. generate a large set of observations from the model to eliminate sampling variability
3. for each parameter, vary its value  around  the true while holding the other parameters are their true values, and evaluate the likelihood of the data
4. for each parameter in 3, record the value that maximizes the log likelihood

The values that maximize the log likelihood should be close to the values you selected in 1. You can also plot the values to ensure that the log likelihood is sensitive to changes in the value of each parameter.  If the curves are relatively flat, the sampler may fail to converge. This will at least rule out a problem with the model. 

The next step after that would be to find the simplest model that reproduces these problems. In principle, it should be easier to identify the problem with a MWE.

**Comment by itsdfish** [comment]
Just a quick update:

I think this might be a scaling issue. I am able to replicate the behavior with the simple example below. I found similar results with BayesianTool.r. It might be worth asking the Turing developers for advice on reparameterization. Sometimes you can rescale or modify the parameters. In other cases, you can rescale the raw data. 

Another thing to consider is changing the particle updating function:

```
import DifferentialEvolutionMCMC: select_base, recombination!

function my_random_gamma(de, Pt, group)
    Np = length(Pt.Θ)
    # select the base particle θb
    Pb = select_base(group)
    # group without Pb
    group_diff = setdiff(group, [Pt])
    # sample particles for θm and θn
    Pm,Pn = de.sample(de, group_diff, 2, false) 
    # sample gamma weights
    γ₁ = rand(Uniform(.01, .1)) #rescaled
    # set γ₂ = 0 after burnin
    γ₂ = de.iter > de.burnin ? 0.0 : rand(Uniform(.01, .1)) # rescaled
    # sample noise for each parameter
    b = Uniform(-de.ϵ, de.ϵ)
    # compute proposal value
    Θp = Pt + γ₁ * (Pm - Pn) + γ₂ * (Pb - Pt) + b
    # reset each parameter to previous value with probability (1-κ)
    recombination!(de, Pt, Θp)
    return Θp
end
```

I rescaled the gamma parameters (you may need to consider other values). You can pass the function to DE with `generate_proposal=my_random_gamma`.  The goal with this approach is to scale down the step size. I'm not sure if it will work, but it might be worth a try. 

Here is a MWE of the convergence problem:

```
cd(@__DIR__)
using DifferentialEvolutionMCMC, Random, Distributions

Random.seed!(50514)

priors = (
    μ = (Normal(0, 10),),
    σ = (Truncated(Cauchy(0, 1), 0.0, Inf),)
)

bounds = ((-Inf,Inf),(0.0,Inf))

data = rand(Normal(0.0, .0001), 50)

function loglike(data, μ, σ)
    return sum(logpdf.(Normal(μ, σ), data))
end

model = DEModel(; priors, model=loglike, data)

de = DE(;bounds, burnin=1000, priors)
n_iter = 2000
chains = sample(model, de, MCMCThreads(), n_iter, progress=true)
```

**Comment by fkrauer** [comment]
Thanks a lot @itsdfish . I am doing a trial with Turing (NUTS) atm, but it is not going well at all as you have seen. I'll try with an adapted random gamma proposal later this week as you suggested. I already tried to rescale all params between 0 and 1, but it didn't change much. Will keep you updated.

**Comment by fkrauer** [comment]
So I have tried different versions, with all parameters rescaled between 0 and 1, as well as at the original scale, and different combinations of sampler parameters and your suggested random gamma proposal. The best I could come up with is the result below, which is still poor mixing for rho0. This was achieved with `θsnooker=0.1, α=0.1, β=0.2, ϵ=0.001, σ=0.1 κ=0.5` and 1'500'000 iterations (1'200'000 burnin) and parameters at original scale. Gelmandiag() suggests all params are converged except rho0. I am not sure running for longer will solve this issue to be honest. Do you have any recommendation how to save this model fit? Thanks a lot. 
![trace_035](https://user-images.githubusercontent.com/36159611/123243044-ec794500-d4e2-11eb-9064-25bff505479f.png)

**Comment by itsdfish** [comment]
I was hoping that the Turing developers would provide some more insight. Considering that you reparameterized the model, I wonder if it is identifiable. Sometimes the correlation matrix of the posterior distribution provides some evidence of non identifiability. Are some of the correlations close to |r|  = 1 ? 

Destructuring the chains into a DataFrame and saving it as CSV file seems like the simplest approach. But I am not sure whether it would be easy to read in such a big file.  Another simple approach might be a [.jld file](https://github.com/JuliaIO/JLD.jl), which is supposed to save a copy of your Julia workspace.

**Comment by fkrauer** [comment]
I have changed the likelihood function from Poisson to Negative Binomial (with estimated dispersion param psi), and it improved the mixing for the rho parameter a bit. I'll run this chain for a bit longer (this one was 500000) and maybe adjust the params of the DEMCMC algorithm a bit to see if I can improve the mixing further. 
![trace_043](https://user-images.githubusercontent.com/36159611/124012360-5f9e2200-d9e1-11eb-9a42-b3fbc5ad11c9.png)

**Comment by itsdfish** [comment]
Unfortunately, it looks like the sampling of the other parameters is less efficient now. In the previous plot, phi was much better. 

Did you try bringing up the issue in the Turing slack channel?

**Comment by fkrauer** [comment]
That's with the default settings for the DEMCMC (k=1.0, beta=0.1). The first plot was with k=0.5, beta=0.2. I'll see if I can improve this by tweaking beta and k a bit. 

I opened an issue in the Turing Github Repo, but haven't received a reply yet. I think I'll post a question on the julia lang forum about Turing. There are probably other people having problems with it, too. But if I can get a good mixing with DEMCMC, I'll go with your library. I like the setup better anyway, with the explicit definition of the loglik function. The Turing DSL is difficult to debug.

**Comment by itsdfish** [comment]
In your post on the forum, you might mention that you are having difficulty with this algorithm too. I'm quite surprised that it is not performing more efficiently after reparameterizing. Hopefully someone will have some insights.

**Comment by itsdfish** [comment]
After looking over your model again, one thing that seems suspicious is the data generating process, which uses uniformly distributed error.

```
data = round.(incidence .* rand(Uniform(0.5,1.5), size(incidence)))
```

However, your likelihood function uses a Poisson PDF. I wouldn't be surprised if that is causing problems.

**Comment by itsdfish** [comment]
Please review the code below for accuracy. It plots the log likelihood as a function of a given parameter while others are held constant at the data-generating values.  The log likelihood surfaces appear to be somewhat irregular. In each case, the peak is far away from the data-generating value (in red). For some parameters, the log likelihood is -Inf at certain boundaries.  The surface for omega appears to be periodic and decreasing, which might cause some problems.  Assuming there are no errors, I would expect most samplers to fail with this model. However, something seems off, either with the code or the model itself. 

```
using DifferentialEquations, StatsBase, StatsPlots
using DifferentialEvolutionMCMC, Distributions, Random, StatsBase, LabelledArrays

Random.seed!(42)

# ODE model: simple SIR model with seasonally forced contact rate
function SIR!(du,u,p,t)

    # states
    S, I, R, C = u
    N = S + I + R
  
    # params
    β = p.β
    η = p.η
    φ = p.φ
    ω = p.ω
    μ = p.μ
    σ = p.σ
      
    # FOI
    βeff = β * (1.0+η*sin(2.0*π*(t-φ)/365.0))
    λ = βeff*I/N

    # change in states
    du[1] = μ*N - λ*S - μ*S + ω*R
    du[2] = λ*S - σ*I - μ*I
    du[3] = σ*I - μ*R - ω*R
    du[4] = σ*I # cumulative incidence
  
end
  
# Calculate incident cases from cumulative incidence
function get_incidence(sol)
    incidence = [sol.u[t].C - sol.u[t-1].C  for t = 2:length(sol.u)]
    return incidence
end

# Solver settings
tmin = 0.0
tmax = 10.0*365.0
tspan = (tmin, tmax)
abstol = 1.0e-10 
reltol = 1.0e-8 
maxiters = 1e7 
saveat = 1.0
solver = AutoVern7(Rodas5())

# Initiate ODE problem
p = @LArray [0.28,0.07,20,1.0/365,1.0/(80*365),1.0/5.0] (:β,:η,:φ,:ω,:μ,:σ) 
u0 = @LArray [9999.0,1.0,0.0,1.0] (:S,:I,:R,:C)

# ODE problem and solution
problem = ODEProblem(SIR!,u0,tspan,p)
sol = solve(problem, solver, abstol=abstol, reltol=reltol, maxiters=maxiters, saveat=saveat)

incidence =get_incidence(sol)


# Fake some data from model
data = @. rand(Poisson(incidence))

# Log Likelihood function
function loglik(data, solver, problem, θ...)

    β, η, ω, φ = θ
    p_new = @LArray [β,η,ω,φ,1.0/(80.0*365.0),1.0/4.98] (:β,:η,:ω,:φ,:μ,:σ)
    
    # Update problem and solve ODEs
    problem_new = remake(problem, p=p_new)
    sol_new = solve(problem_new, solver, abstol=1.0e-10 , reltol=1.0e-8 , maxiters=1e7, saveat=1)
 
    # calculate incidence
    incidence = get_incidence(sol_new)

    # due to numerical instability of the diff eqs, some states can become negative. In this case, return -Inf
    if minimum(incidence) < 0.0
        ll = -Inf
    else
        ll = sum(logpdf.(Poisson.(incidence), data))
    end
    return ll
end

pyplot()
# -Inf below .2
x = .1:.01:.5
LLs = map(x-> loglik(data, solver, problem, x, 0.07,20,1.0/365), x)
plot(x, LLs, xlabel="β", ylabel="loglikelihood", grid=false, leg=false)
vline!([.28], linestyle=:dash)

# -Inf after 1.5
x = .01:.05:2.0
LLs = map(x-> loglik(data, solver, problem, .28, x, 20,1.0/365), x)
plot(x, LLs, xlabel="η", ylabel="loglikelihood", grid=false, leg=false)
vline!([.07], linestyle=:dash)

x = 1.0:50.0
LLs = map(x-> loglik(data, solver, problem, .28, .07, x, 1.0/365), x)
plot(x, LLs, xlabel="φ", ylabel="loglikelihood", grid=false, leg=false)
vline!([20.0], linestyle=:dash)

# oscolates 
x = .00005:.0001:.005
LLs = map(x-> loglik(data, solver, problem, .28, .07, 20, x), x)
plot(x, LLs, xlabel="ω", ylabel="loglikelihood", grid=false, leg=false)
vline!([1/365], linestyle=:dash)
```

**Comment by fkrauer** [comment]
Thanks a lot for bringing this up, this is indeed odd. I think I have found the culprit: the problem is `θ...` in the loglik function. Somehow it changes the return value of the function.  

The following code returns the correct loglik value (checked with R):

```
function loglik(data, solver, problem, p) 

    # Update problem and solve ODEs
    problem_new = remake(problem, p=p)
    sol_new = solve(problem_new, 
                    solver, 
                    abstol=1.0e-10 , 
                    reltol=1.0e-8 , 
                    isoutofdomain=(u,p,t)->any(x->x<0.0,u), 
                    saveat=1.0)
 
    # calculate incidence
    incidence = get_incidence(sol_new)
    # replace negative incidence with small positive value. This is a numerical instability issue resulting from the ODEs.
    # negative incidence will break the loglik function, so we have to do this
    incidence = ifelse.(incidence .< 0.0, 0.0, incidence)

    ll = sum(a -> sum(a), logpdf.(Poisson.(incidence), data))
    #ll = sum(logpdf.(Poisson.(incidence), data)) # this is equivalent to the line above, but more allocation and slower

    return ll
end
```

for `p = @LArray [0.28,0.07,20.0,1.0/365.0,1.0/(80*365),1.0/5.0] (:β,:η,:φ,:ω,:μ,:σ)` it returns a loglik value around --6856. 

However, the following code, returns a completely different result:

```
function loglik(data, solver, problem, θ...) 

    β, η, ω, φ = θ
    p = @LArray [β,η,φ,ω,1.0/(80.0*365.0),1.0/5.0] (:β,:η,:φ,:ω,:μ,:σ)

    # Update problem and solve ODEs
    problem_new = remake(problem, p=p)
    sol_new = solve(problem_new, 
                    solver, 
                    abstol=1.0e-10 , 
                    reltol=1.0e-8 , 
                    isoutofdomain=(u,p,t)->any(x->x<0.0,u), 
                    saveat=1.0)
 
    # calculate incidence
    incidence = get_incidence(sol_new)
    # replace negative incidence with small positive value. This is a numerical instability issue resulting from the ODEs.
    # negative incidence will break the loglik function, so we have to do this
    incidence = ifelse.(incidence .< 0.0, 0.0, incidence)

    ll = sum(a -> sum(a), logpdf.(Poisson.(incidence), data))
    #ll = sum(logpdf.(Poisson.(incidence), data)) # this is equivalent to the line above, but more allocation and slower

    return ll
end
```

If I test this with `loglik(data, solver, problem, p[1:4]...)` the result is around -1.8716225757315452e6. 

The reason for the `θ...` was to be able to have the estimated parameters as one argument in the loglik function and the function for the sampler, without writing out all the parameters explicitly. Is there a way around this problem?

**Comment by itsdfish** [comment]
I think I see the problem. In the last code I posted, the parameters are ordered as follows when generating the data: β, η, φ ,ω, 

See the following:

`p = @LArray [0.28,0.07,20,1.0/365,1.0/(80*365),1.0/5.0] (:β,:η,:φ,:ω,:μ,:σ) `

Inside your loglik function, omega and phi are switched:
    
`β, η, ω, φ = θ`

So all you need to do is make sure the parameters are ordered consistently throughout.  As noted in the README, the priors must be specified in the same order also. That should fix your model.

**Comment by fkrauer** [comment]
yes, I just realized this as well, was about to post it :). I have corrected it, and the loglik profiles look totally fine now. Thanks a lot for bringing this up. I'll fit this and post the results. 
![omega](https://user-images.githubusercontent.com/36159611/124141852-06d69400-da8a-11eb-9fd4-6770d392ee2e.png)

**Comment by itsdfish** [comment]
No problem. I'm glad we were able to eventually figure out the problem. What you have posted above looks much better.  I always generate those plots to double check the log likelihood function. I also add unit tests to check that the max likelihood is close to the data-generating value when a large sample size is taken. 

I wanted to use keyword arguments for the loglike function, but I had trouble making it work. If that could be accomplished, you would not have an error because:

```
julia> using Parameters

julia> x = (a=3, b=1)
(a = 3, b = 1)

julia> @unpack b,a = x
(a = 3, b = 1)

julia> a
3

julia> b
1
```
Unfortunately, the current setup creates tuple which is order-dependent.

**Comment by fkrauer** [comment]
It worked perfectly! 500,000 iters (300,000 burnin). Amazing. Thanks a lot for your help. This was a toy model, so now I'll move on to the real model, which is a bit more complex. Fingers crossed that I get equally good mixing....

![trace_toy](https://user-images.githubusercontent.com/36159611/124164000-8707f400-daa0-11eb-8bf6-7b97660e8cfd.png)

**Comment by itsdfish** [comment]
Looks good!

Just for future reference, did you use the standard snooker setup, or did you modify the setup somehow?

**Comment by fkrauer** [comment]
I used the snooker update with the default theta_snooker = 0.1 and random_gamma proposal. 

Here is the code if needed:
```
# Household stuff
cd(@__DIR__)
using Pkg 
Pkg.activate(".") 


using DifferentialEquations, StatsBase, StatsPlots
using DifferentialEvolutionMCMC, Distributions, Random, StatsBase, LabelledArrays
using MCMCChains

Random.seed!(42)

# ODE model: simple SIR model with seasonally forced contact rate
function SIR!(du,u,p,t)

    # states
    S, I, R, C = u
    N = S + I + R
  
    # params
    β = p.β
    η = p.η
    φ = p.φ
    ω = 1/p.ω
    μ = p.μ
    σ = p.σ
      
    # FOI
    βeff = β * (1.0+η*cos(2.0*π*(t-φ)/365.0))
    λ = βeff*I/N

    # change in states
    du[1] = μ*N - λ*S - μ*S + ω*R
    du[2] = λ*S - σ*I - μ*I
    du[3] = σ*I - μ*R - ω*R
    du[4] = σ*I # cumulative incidence
  
end
  
# Calculate incident cases from cumulative incidence
function get_incidence(sol)
    incidence = [sol.u[t].C - sol.u[t-1].C  for t = 2:length(sol.u)]
    return incidence
end

# Solver settings
tmin = 0.0
tmax = 10.0*365.0
tspan = (tmin, tmax)
abstol = 1.0e-3
reltol = 1.0e-3 
saveat = 7.0
solver = AutoTsit5(Rosenbrock23())

# Initiate ODE problem
const theta_fix = [1.0/(80*365),1.0/5.0]
theta_est = [0.28,0.07,20.0,365.0]
parnames =  (:β,:η,:φ,:ω,:μ,:σ) 
p = @LArray [theta_est; theta_fix] parnames
const u0 = @LArray [9999.0,1.0,0.0,1.0] (:S,:I,:R,:C)

# ODE problem and solution
problem = ODEProblem(SIR!,u0,tspan,p)
sol = solve(problem, solver, abstol=abstol, reltol=reltol, saveat=saveat)

incidence = get_incidence(sol)
incidence = ifelse.(incidence .< 0.0, 0.0, incidence)

plot(incidence)

# Fake some data from model
data = @. rand(Poisson(incidence))

# Log Likelihood function
function loglik(data, solver, problem, theta_fix, parnames, theta_est...) 

    p_new = @LArray [theta_est...; theta_fix] parnames

    # Update problem and solve ODEs
    problem_new = remake(problem, p=p_new)
    sol_new = solve(problem_new, 
                    solver, 
                    abstol=1.0e-3, 
                    reltol=1.0e-3, 
                    isoutofdomain=(u,p,t)->any(x->x<0.0,u), 
                    saveat=7.0)
 
    # calculate incidence
    incidence = get_incidence(sol_new)
    # replace negative incidence with small positive value. This is a numerical instability issue resulting from the ODEs.
    # negative incidence will break the loglik function, so we have to do this
    incidence = ifelse.(incidence .< 0.0, 0.0, incidence)

    if length(incidence) < length(data)
    ll = -Inf
    else
    ll = sum(a -> sum(a), logpdf.(Poisson.(incidence), data))
    #ll = sum(logpdf.(Poisson.(incidence), data)) # both works

    end

    return ll

end

# check function
loglik(data, solver, problem, theta_fix, parnames, theta_est...)

#=
#pyplot()
# -Inf below .2
x = 0.1:.01:1.0
LLs = map(x-> loglik(data, solver, problem, x, 0.07,20.0,1.0/365), x)
plot(x, LLs, xlabel="β", ylabel="loglikelihood", grid=false, leg=false)
vline!([.28], linestyle=:dash)

# -Inf after 1.5
x = .01:.05:2.0
LLs = map(x-> loglik(data, solver, problem, .28, x, 20.0,1.0/365), x)
plot(x, LLs, xlabel="η", ylabel="loglikelihood", grid=false, leg=false)
vline!([.07], linestyle=:dash)

x = 0.0:364.0
LLs = map(x-> loglik(data, solver, problem, .28, .07, x, 1.0/365), x)
plot(x, LLs, xlabel="φ", ylabel="loglikelihood", grid=false, leg=false)
vline!([20.0], linestyle=:dash)

# oscolates 
x = 0.0:.0001:1.0
LLs = map(x-> loglik(data, solver, problem, .28, .07, 20.0, x), x)
plot(x, LLs, xlabel="ω", ylabel="loglikelihood", grid=false, leg=false)
vline!([1/365], linestyle=:dash)

=#

# Priors
const priors = (
    β = (Uniform(0.0,1.0),), 
    η = (Uniform(0.0,1.0),),
    φ = (Uniform(0.0,364.0),),
    ω = (Uniform(1.0, 5.0*365.0),)
)    

keys(priors)==parnames[1:length(theta_est)] # must be true

# Upper and lower bounds (is equivalent to argument of the Uniform)
const bounds = (
    (0.0,1.0), #β
    (0.0,1.0), # η
    (0.0,364.0), # φ
    (1.0, 5.0*365.0) #ω
)

const burnin= 300000
const n_iter = 500000
const n_groups = 3 ## number of groups of particles. default is 4
const Np = 3 #*length(theta_est) # N of particles/chains per group. Default is nparam*3 for DE, and 3 for DEzs
const n_initial=10*length(theta_est)
demodel = DEModel(solver, problem, theta_fix, parnames; priors, model=loglik, data)
de = DE(; n_groups, priors, Np, discard_burnin=false, sample=resample, n_initial, θsnooker=0.1, burnin, bounds) 
@time trace = sample(demodel, de, MCMCThreads(), n_iter, progress=true)

```

**Comment by itsdfish** [comment]
Much appreciated!

**Comment by fkrauer** [comment]
I have tried to fit a more complex model (17 compartments) but it has always resulted in very poor mixing despite tuning the algorithm parameters and the proposal function. I could get an acceptable mixing (at least for some params) when fitting to data simulated from the model itself (see traceplots below). Generally, for this type of model and loglikelihood landscape Gibbs updating (k=0.0) gave the best mixing when fitting to real data, but I couldn't get full convergence. I might have to run it longer and rescale the parameter, but I haven't found a good way to rescale them. I have decided to go with Turing for this specific model. The mixing, convergence and the posteriors are very good so far. The params are automatically converted to real space, so no need to think about this. I will keep your package in mind for simpler models, I think it can definitely work if the model is well specified and the likelihood tractable. Thanks for all the help, I appreciate it. 

trace after removing burnin
![trace_061](https://user-images.githubusercontent.com/36159611/128561498-4a723198-68d1-44a8-a26d-139b64115e9e.png)

full trace with burnin
![trace_raw_062](https://user-images.githubusercontent.com/36159611/128561501-db444a66-a4dc-4f19-8e5c-ddd6ef6bc157.png)
